Remove dead/unreferenced data in ROMClasses

AlexeyKhrabrov opened the pull request:
This pull request fixes the following issues in the ROMClass builder that prevented ROMClass serialization in JITServer (#11331) from working correctly.
- The total size of the ROMClass (the `romSize` field) was calculated incorrectly: it included the size of the variable handle method type lookup table twice. As a result, many ROMClasses on jdk11+ include an unused zero-filled range at the end.
- When debug info is excluded from a ROMClass (e.g. with `-Xnolinenumbers`), the source file name string was still included in the UTF8 section, even though it's not referenced anywhere in the ROMClass.

AlexeyKhrabrov commented:
@tajila Hello Tobi, can I please get a review for this PR? @mpirvu FYI this is a prerequisite for #11331.

tajila commented:
@babsingh please review these changes

babsingh reviewed (commented):
### Code removed
1. `ROMClassBuilder.hpp`: `SizeInformation.varHandleMethodTypeLookupTableSize` field
2. `ROMClassWriter.hpp`: `getVarHandleMethodTypePaddedSize` method

**_Why is `varHandleMethodTypeLookupTableSize` tagged as dead code?_**

### Background info

The code being removed is used to write the `VarHandle MethodType lookup table (and its size)` to the `J9ROMClass`.

### Other dependent code

**_Will the below code work if the above mentioned code is removed?_**

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L312

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L413

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L438

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L821-L842

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L1868-L1877

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/oti/VMHelpers.hpp#L1451-L1469

tajila commented:
Jenkins test sanity,extended win jdk8

babsingh commented:
@tajila JDK8 does not have VarHandles. We will need to re-run the test jobs with JDK11+ to perform a functional verification for these changes.

tajila commented:
Jenkins test sanity,extended xlinux jdk11

AlexeyKhrabrov commented:
@babsingh In response to your comments/questions and to better explain what the VarHandle fix is about:

The `SizeInformation::varHandleMethodTypeLookupTableSize` field and the `ROMClassWriter::getVarHandleMethodTypePaddedSize()` method are only ever used to calculate the total `romSize`. They become dead code (not used anywhere) after removing the addition of the table size from those calculations.

The size of the VarHandle MethodType lookup table is already included in `rcWithOutUTF8sSize` here:

https://github.com/eclipse/openj9/blob/4b096cb503f150765de2bf20b9e17fedc649ca6d/runtime/bcutil/ROMClassBuilder.cpp#L420

It is already accounted for during the "1st phase" of ROMClass construction (when this code is called with `markAndCountOnly == true`):

https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L1868-L1877
https://github.com/eclipse/openj9/blob/2dec40f6478d2f4fe8f754c5058bd174d34513f4/runtime/bcutil/ROMClassWriter.cpp#L709-L722

babsingh commented:
@AlexeyKhrabrov https://github.com/eclipse/openj9/pull/11566#issuecomment-754855484 clarifies as well as justifies the reasoning for this change. These changes look functionally good to me.

I would suggest ...
1. Squashing commits, `Remove dead code in ROMClass size info calculation` and `Fix the value of the romSize field in ROMClass`, into one commit since the removal of the `varHandleMethodTypeLookupTableSize` field and its usage are dependent.
2. Adding a comment in `ROMClassWriter::writeVarHandleMethodTypeLookupTable` to highlight the `inclusion of the size of the VarHandle MethodType lookup table` in the same commit.
3. Updating the commit message with the key details from https://github.com/eclipse/openj9/pull/11566#issuecomment-754855484.

The above changes will improve the documentation for future reference. 

Tentatively approving the changes on successful improvement of documentation. Passing it to @tajila for final confirmation and merge.

No code changes have been suggested during the review. So, we can reuse the results of the current PR jobs:
- JDK8 build: https://ci.eclipse.org/openj9/job/Build_JDK8_x86-64_windows_Personal/1020/
- JDK8 sanity: https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal/621/
- JDK8 extended: https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_extended.functional_x86-64_windows_Personal/153/
- JDK11 build: https://ci.eclipse.org/openj9/job/Build_JDK11_x86-64_linux_Personal/1082/
- JDK11 sanity: https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal/750/
- JDK11 extended: https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_extended.functional_x86-64_linux_Personal/86/

AlexeyKhrabrov commented:
I updated the commits and their messages as suggested, and added the comments about the size of the table already being included in `romSize` in `ROMClassBuilder.cpp` (I think it's a better place for them).
